rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,395 @@
+and does not violate our kube cluster name validation regex `^[a-zA-Z0-9._-]+$`.
+
+#### Azure Discovery
+
+Azure resources have a resource ID that uniquely identifies the resource, e.g.:
+`/subscriptions/00000000-1111-2222-3333-444444444444/resourceGroups/<group name>/providers/<provider name>/<name>`
+
+We could use this ID as the database name, but it is unnecessarily verbose.
+It will also fail to match our database name validation regex:
+`[a-z]([-a-z0-9]*[a-z0-9])?`.
+
+Additionally, all of the Azure databases that Teleport currently supports
+require globally unique names (within the same type of database), because Azure
+assigns a DNS name:
+
+- Redis: `<name>.redis.cache.windows.net`.
+- SQL Server: `<name>.database.windows.net`.
+- Postgres: `<name.postgres.database.azure.com`.
+- MySQL: `<name>.mysql.database.azure.com`.
+
+MySQL/Postgres server names must be unique among both single-server and
+flexible-server instances.
+
+Therefore, we can form a uniquely identifying name among Azure resources just by
+adding the kind of matcher to the resource name.
+However, AKS kube clusters do not require globally unique names - they only need
+to be unique within the same resource group in the same subscription.
+
+To make the naming convention consistent, and to "future-proof" it, the
+naming convention will be to append a suffix that includes:
+
+- Name of the Azure matcher type
+  - `aks`, `mysql`, `postgres`, `redis`, `sqlserver` (as of writing this RFD)
+- Azure resource group name
+  - resource group names may contain characters that we do not allow in database
+    or kube cluster names.
+    The resource group name should be checked for invalid characters and dropped
+    from the name suffix if it is invalid.
+    This is only a heuristic, but any approach here will be a heuristic, and
+    this is the simplest string transform we can do, which avoids confusing
+    users with strange resource group names they don't recognize.
+- Azure subscription ID
+  - subscription IDs only contains letters, digits, and hyphens.
+
+Example:
+```yaml
+discovery_service:
+  enabled: true
+  aws:
+    - types: ["aks", "mysql", "postgres"]
+      regions: ["eastus"]
+      subscriptions:
+        - "11111111-1111-1111-1111-111111111111"
+        - "22222222-2222-2222-2222-222222222222"
+      resource_groups: ["group1", "group2", "weird-)(-group-name"]
+      tags:
+        "*": "*"
+```
+
+If the discovery service is configured like the above, the discovery agent will
+discover Azure AKS kube clusters, Azure MySQL, and Azure PostgreSQL databases.
+
+Now suppose that four AKS kube clusters named `foo` exist in each combination of
+resource group and subscription ID, and a MySQL database and Postgres database
+both named `foo` exist in the the `1111..` subscription and `group1`.
+If the discovery service applies the new naming convention, the discovered
+resources should be named:
+
+- `foo-aks-group1-11111111-1111-1111-1111-111111111111`
```

Comment:
Would it make sense to include region as well, for consistency in how we're generating names between different cloud providers?

Reply:
We could do that for consistency, but it doesn't provide any benefit for avoiding name collisions, since AKS names are unique within a resource group + subscription, i.e. the only way you can have `foo-aks-eastus-...` and `foo-aks-westus-...` is if the two are in different resource group and/or subscription. So it would look like:

```
foo-aks-eastus-group1-11111111-1111-1111-1111-111111111111
foo-aks-westus-group2-22222222-2222-2222-2222-222222222222
```

Reply:
might be beneficial though, if the resource group name contains invalid chars that requires we do a string transform on it as I detailed

Reply:
7b21853436b13ddfc639101d00aaf64d76664d39